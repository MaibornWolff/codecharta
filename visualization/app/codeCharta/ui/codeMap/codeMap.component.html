<div id="codeMap" ng-hide="$ctrl._viewModel.isLoadingFile">
	<view-cube-component ng-class="{ sideBarVisible: $ctrl._viewModel.isSideBarVisible }"></view-cube-component>
	<attribute-side-bar-component></attribute-side-bar-component>
</div>
