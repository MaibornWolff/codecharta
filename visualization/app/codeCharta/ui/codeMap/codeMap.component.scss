code-map-component {
	view-cube-component {
		position: absolute;
		z-index: 11;
		right: 0px;
		top: 10px;

		&.sideBarVisible {
			right: 350px;
		}
	}
}
