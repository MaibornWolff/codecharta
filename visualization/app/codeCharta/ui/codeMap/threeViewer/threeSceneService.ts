import { AmbientLight, DirectionalLight, Scene, Group } from "three"
import { CodeMapMesh } from "../rendering/codeMapMesh"
import { CodeMapBuilding } from "../rendering/codeMapBuilding"
import { CodeMapPreRenderServiceSubscriber, CodeMapPreRenderService } from "../codeMap.preRender.service"
import { IRootScopeService } from "angular"
import { StoreService } from "../../../state/store.service"

export interface BuildingSelectedEventSubscriber {
	onBuildingSelected(selectedBuilding?: CodeMapBuilding)
}

export interface BuildingDeselectedEventSubscriber {
	onBuildingDeselected()
}

export interface CodeMapMeshChangedSubscriber {
	onCodeMapMeshChanged(mapMesh: CodeMapMesh)
}

export class ThreeSceneService implements CodeMapPreRenderServiceSubscriber {
	private static readonly BUILDING_SELECTED_EVENT = "building-selected"
	private static readonly BUILDING_DESELECTED_EVENT = "building-deselected"
	private static readonly CODE_MAP_MESH_CHANGED_EVENT = "code-map-mesh-changed"

	scene: Scene
	labels: Group
	edgeArrows: Group
	mapGeometry: Group
	private readonly lights: Group
	private mapMesh: CodeMapMesh

	private selected: CodeMapBuilding = null
	private highlighted: CodeMapBuilding[] = []

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
		CodeMapPreRenderService.subscribe(this.$rootScope, this)

		this.scene = new Scene()
		this.mapGeometry = new Group()
		this.lights = new Group()
		this.labels = new Group()
		this.edgeArrows = new Group()

		this.initLights()

		this.scene.add(this.mapGeometry)
		this.scene.add(this.edgeArrows)
		this.scene.add(this.labels)
		this.scene.add(this.lights)
	}

	onRenderMapChanged() {
		this.reselectBuilding()
	}

	highlightBuildings() {
		const state = this.storeService.getState()
		this.getMapMesh().highlightBuilding(this.highlighted, this.selected, state)
	}

	highlightSingleBuilding(building: CodeMapBuilding) {
		this.highlighted = []
		this.addBuildingToHighlightingList(building)
		this.highlightBuildings()
	}

	addBuildingToHighlightingList(building: CodeMapBuilding) {
		this.highlighted.push(building)
	}

	clearHighlight() {
		if (this.getMapMesh()) {
			this.getMapMesh().clearHighlight(this.selected)
			this.highlighted = []
		}
	}

	selectBuilding(building: CodeMapBuilding) {
		const color = this.storeService.getState().appSettings.mapColors.selected
		this.getMapMesh().selectBuilding(building, color)
		this.selected = building
		this.highlightBuildings()
		this.$rootScope.$broadcast(ThreeSceneService.BUILDING_SELECTED_EVENT, this.selected)
	}

	clearSelection() {
		if (this.selected) {
			this.getMapMesh().clearSelection(this.selected)
			this.$rootScope.$broadcast(ThreeSceneService.BUILDING_DESELECTED_EVENT)
		}
		if (this.highlighted.length > 0) {
			this.highlightBuildings()
		}
		this.selected = null
	}

	initLights() {
		const ambilight = new AmbientLight(0x707070) // soft white light
		const light1 = new DirectionalLight(0xe0e0e0, 1)
		light1.position.set(50, 10, 8).normalize()

		const light2 = new DirectionalLight(0xe0e0e0, 1)
		light2.position.set(-50, 10, -8).normalize()

		this.lights.add(ambilight)
		this.lights.add(light1)
		this.lights.add(light2)
	}

	setMapMesh(mesh: CodeMapMesh) {
		const { mapSize } = this.storeService.getState().treeMap
		this.mapMesh = mesh

		while (this.mapGeometry.children.length > 0) {
			this.mapGeometry.remove(this.mapGeometry.children[0])
		}

		this.mapGeometry.position.x = -mapSize
		this.mapGeometry.position.y = 0
		this.mapGeometry.position.z = -mapSize

		this.mapGeometry.add(this.mapMesh.getThreeMesh())
		this.notifyMapMeshChanged()
	}

	getMapMesh() {
		return this.mapMesh
	}

	scale() {
		const { mapSize } = this.storeService.getState().treeMap
		const scale = this.storeService.getState().appSettings.scaling

		this.mapGeometry.scale.set(scale.x, scale.y, scale.z)
		this.mapGeometry.position.set(-mapSize * scale.x, 0, -mapSize * scale.z)
		this.mapMesh.setScale(scale)
	}

	getSelectedBuilding() {
		return this.selected
	}

	getHighlightedBuilding() {
		return this.highlighted[0]
	}

	getHighlightedNode() {
		if (this.getHighlightedBuilding()) {
			return this.getHighlightedBuilding().node
		}
		return null
	}

	private reselectBuilding() {
		if (this.selected) {
			const buildingToSelect: CodeMapBuilding = this.getMapMesh().getBuildingByPath(this.selected.node.path)
			if (buildingToSelect) {
				this.selectBuilding(buildingToSelect)
			}
		}
	}

	private notifyMapMeshChanged() {
		this.$rootScope.$broadcast(ThreeSceneService.CODE_MAP_MESH_CHANGED_EVENT, this.mapMesh)
	}

	static subscribeToBuildingDeselectedEvents($rootScope: IRootScopeService, subscriber: BuildingDeselectedEventSubscriber) {
		$rootScope.$on(this.BUILDING_DESELECTED_EVENT, () => {
			subscriber.onBuildingDeselected()
		})
	}

	static subscribeToBuildingSelectedEvents($rootScope: IRootScopeService, subscriber: BuildingSelectedEventSubscriber) {
		$rootScope.$on(this.BUILDING_SELECTED_EVENT, (_event, selectedBuilding: CodeMapBuilding) => {
			subscriber.onBuildingSelected(selectedBuilding)
		})
	}

	static subscribeToCodeMapMeshChangedEvent($rootScope: IRootScopeService, subscriber: CodeMapMeshChangedSubscriber) {
		$rootScope.$on(this.CODE_MAP_MESH_CHANGED_EVENT, (_event, mapMesh: CodeMapMesh) => {
			subscriber.onCodeMapMeshChanged(mapMesh)
		})
	}
}
