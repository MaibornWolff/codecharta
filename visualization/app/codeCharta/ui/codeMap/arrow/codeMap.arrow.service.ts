import { Injectable, OnDestroy } from "@angular/core"
import { ThreeSceneService } from "../threeViewer/threeSceneService"
import { Node, EdgeVisibility, CcState } from "../../../codeCharta.model"
import { ArrowHelper, BufferGeometry, CubicBezierCurve3, Line, LineBasicMaterial, Object3D, Vector3 } from "three"
import { ColorConverter } from "../../../util/color/colorConverter"
import { CodeMapBuilding } from "../rendering/codeMapBuilding"
import { IdToBuildingService } from "../../../services/idToBuilding/idToBuilding.service"
import { tap } from "rxjs"
import { hoveredNodeIdSelector } from "../../../state/store/appStatus/hoveredNodeId/hoveredNodeId.selector"
import { debounce } from "../../../util/debounce"
import { Store, State } from "@ngrx/store"
import { edgeVisibilitySelector } from "./utils/edgeVisibility.selector"

@Injectable({ providedIn: "root" })
export class CodeMapArrowService implements OnDestroy {
    private map: Map<string, Node>
    private VERTICES_PER_LINE = 5
    private arrows: Object3D[] = new Array<Object3D>()
    private HIGHLIGHT_BUILDING_DELAY = 1
    private debounceCalculation: (hoveredBuilding: CodeMapBuilding) => void = debounce(
        (hoveredBuilding: CodeMapBuilding) => this.resetEdgesOfBuildings(hoveredBuilding),
        this.HIGHLIGHT_BUILDING_DELAY
    )
    private subscription = this.store
        .select(hoveredNodeIdSelector)
        .pipe(
            tap(hoveredNodeId => {
                if (hoveredNodeId !== null) {
                    const hoveredBuilding = this.idToBuildingService.get(hoveredNodeId)
                    this.onBuildingHovered(hoveredBuilding)
                } else {
                    this.onBuildingUnhovered()
                }
            })
        )
        .subscribe()

    constructor(
        private readonly store: Store<CcState>,
        private readonly state: State<CcState>,
        private threeSceneService: ThreeSceneService,
        private readonly idToBuildingService: IdToBuildingService
    ) {
        this.threeSceneService.subscribe("onBuildingSelected", this.onBuildingSelected)
        this.threeSceneService.subscribe("onBuildingDeselected", this.onBuildingDeselected)
    }

    ngOnDestroy(): void {
        this.subscription.unsubscribe()
    }

    private resetEdgesOfBuildings = (hoveredBuilding: CodeMapBuilding) => {
        if (this.isEdgeApplicableForBuilding(hoveredBuilding)) {
            this.clearArrows()
            this.showEdgesOfBuildings(hoveredBuilding)
        }
        this.scale()
    }

    onBuildingSelected = (data: { building: CodeMapBuilding }) => {
        if (this.isEdgeApplicableForBuilding(data.building)) {
            this.clearArrows()
            this.showEdgesOfBuildings()
        }
        this.scale()
    }

    onBuildingDeselected = () => {
        this.clearArrows()
        this.addEdgePreview()
    }

    onBuildingHovered(hoveredBuilding: CodeMapBuilding) {
        this.debounceCalculation(hoveredBuilding)
    }

    onBuildingUnhovered() {
        const { isEdgeMetricVisible } = this.state.getValue().appSettings

        if (isEdgeMetricVisible) {
            this.clearArrows()
            this.showEdgesOfBuildings()
        }
        this.scale()
    }

    clearArrows() {
        this.arrows = []
        // Remove all children
        this.threeSceneService.edgeArrows.children.length = 0
    }

    addArrow(arrowTargetNode: Node, arrowOriginNode: Node, buildingIsOriginNode: boolean) {
        const { appSettings, dynamicSettings } = this.state.getValue()
        const curveScale = 100 * appSettings.edgeHeight

        if (arrowTargetNode.attributes?.[dynamicSettings.heightMetric] && arrowOriginNode.attributes?.[dynamicSettings.heightMetric]) {
            const curve = this.createCurve(arrowOriginNode, arrowTargetNode, curveScale)
            const color = ColorConverter.getNumber(appSettings.mapColors[buildingIsOriginNode ? "outgoingEdge" : "incomingEdge"])
            this.highlightBuilding(buildingIsOriginNode ? arrowTargetNode : arrowOriginNode)
            this.setCurveColor(curve, color)
        }
    }

    addEdgePreview() {
        const edges = edgeVisibilitySelector(this.state.getValue())

        for (const edge of edges) {
            const originNode = this.map.get(edge.fromNodeName)
            const targetNode = this.map.get(edge.toNodeName)
            if (originNode && targetNode && edge.visible !== EdgeVisibility.none && edge.visible) {
                //TODO It seems originNode or targetNode might be undefined here,
                // I think it results from the method being called multiple times when it might not be available yet
                // I changed that back to avoid console errors and re-enable the edge-metric, however we should investigate why this is happening
                const curveScale = 100 * this.state.getValue().appSettings.edgeHeight
                const curve = this.createCurve(originNode, targetNode, curveScale)
                this.previewMode(curve, edge.visible)
            }
        }
    }

    addEdgeMapBasedOnNodes(nodes: Node[]) {
        this.map = this.getNodesAsMap(nodes)
    }

    scale() {
        const { scaling } = this.state.getValue().appSettings
        for (const arrow of this.arrows) {
            arrow.scale.x = scaling.x
            arrow.scale.y = scaling.y
            arrow.scale.z = scaling.z
        }
    }

    private isEdgeApplicableForBuilding(codeMapBuilding: CodeMapBuilding) {
        return this.state.getValue().appSettings.isEdgeMetricVisible && codeMapBuilding && !codeMapBuilding.node.flat
    }

    private showEdgesOfBuildings(hoveredbuilding?: CodeMapBuilding) {
        const buildings: Map<string, Node> = new Map()
        const selectedBuilding = this.threeSceneService.getSelectedBuilding()

        if (selectedBuilding) {
            const { node } = selectedBuilding
            buildings.set(node.path, node)
        }
        if (hoveredbuilding) {
            const { node } = hoveredbuilding
            buildings.set(node.path, node)
        }
        if (buildings.size > 0) {
            this.buildPairingEdges(buildings)
        } else {
            this.addEdgePreview()
        }
    }

    private buildPairingEdges(node: Map<string, Node>) {
        const { edges } = this.state.getValue().fileSettings

        for (const edge of edges) {
            const originNode = this.map.get(edge.fromNodeName)
            // TODO: Maps should only have valid edges. If that's not the case, the
            // internal decoration is likely faulty. Check if only test data is not
            // correct or what the root cause of these checks actually is.
            if (originNode === undefined) {
                continue
            }
            const targetNode = this.map.get(edge.toNodeName)
            if (targetNode === undefined) {
                continue
            }
            if (node.has(originNode.path)) {
                this.addArrow(targetNode, originNode, true)
                this.threeSceneService.highlightBuildings()
                // TODO: Check if the second if case is actually necessary. Edges should
                // always have valid origin and target paths. The test data is likely
                // faulty and should be improved.
            } else if (node.has(targetNode.path)) {
                this.addArrow(targetNode, originNode, false)
                this.threeSceneService.highlightBuildings()
            }
        }
    }

    private createCurve(arrowOriginNode: Node, arrowTargetNode: Node, curveScale) {
        const bezierPoint2 = arrowOriginNode.outgoingEdgePoint.clone()
        const bezierPoint3 = arrowTargetNode.incomingEdgePoint.clone()
        const arrowHeight = Math.max(bezierPoint2.y + arrowTargetNode.height, bezierPoint3.y + 1) + curveScale
        bezierPoint2.setY(arrowHeight)
        bezierPoint3.setY(arrowHeight)
        return new CubicBezierCurve3(arrowOriginNode.outgoingEdgePoint, bezierPoint2, bezierPoint3, arrowTargetNode.incomingEdgePoint)
    }

    private highlightBuilding(node: Node) {
        const building = this.threeSceneService.getMapMesh().getMeshDescription().getBuildingByPath(node.path)
        this.threeSceneService.addBuildingToHighlightingList(building)
    }

    private setCurveColor(bezier: CubicBezierCurve3, color: number, bezierPoints = 50) {
        const points = bezier.getPoints(bezierPoints)
        const curveObject = this.buildLine(points, color)
        curveObject.add(this.buildArrow(points))
        this.threeSceneService.edgeArrows.add(curveObject)
        this.arrows.push(curveObject)
    }

    private previewMode(curve, edgeVisibility: EdgeVisibility) {
        if (edgeVisibility === EdgeVisibility.both || edgeVisibility === EdgeVisibility.from) {
            const outgoingArrow: Object3D = this.makeArrowFromBezier(curve, false)
            this.threeSceneService.edgeArrows.add(outgoingArrow)
            this.arrows.push(outgoingArrow)
        }

        if (edgeVisibility === EdgeVisibility.both || edgeVisibility === EdgeVisibility.to) {
            const incomingArrow: Object3D = this.makeArrowFromBezier(curve, true)
            this.threeSceneService.edgeArrows.add(incomingArrow)
            this.arrows.push(incomingArrow)
        }
    }

    private getNodesAsMap(nodes: Node[]) {
        const map: Map<string, Node> = new Map()
        for (const node of nodes) {
            map.set(node.path, node)
        }
        return map
    }

    private makeArrowFromBezier(bezier: CubicBezierCurve3, incoming: boolean, bezierPoints = 50) {
        const points = bezier.getPoints(bezierPoints)
        const { incomingEdge, outgoingEdge } = this.state.getValue().appSettings.mapColors
        const arrowColor = incoming ? incomingEdge : outgoingEdge
        const pointsPreviews = incoming
            ? points.slice(bezierPoints + 1 - this.VERTICES_PER_LINE)
            : points.slice(0, points.length - (bezierPoints + 1 - this.VERTICES_PER_LINE))

        return this.buildEdge(pointsPreviews, ColorConverter.getNumber(arrowColor))
    }

    private buildEdge(points: Vector3[], color: number): Object3D {
        const curveObject = this.buildLine(points, color)
        curveObject.add(this.buildArrow(points))

        return curveObject
    }

    private buildLine(points: Vector3[], color = 0) {
        const geometry = new BufferGeometry()
        geometry.setFromPoints(points)

        const material = new LineBasicMaterial({ color, linewidth: 1 })
        return new Line(geometry, material)
    }

    private buildArrow(points: Vector3[], ARROW_COLOR = 0, headLength = 10, headWidth = 10) {
        const direction = points.at(-1).clone().sub(points.at(-2).clone()).normalize()

        const origin = points.at(-1).clone()
        if (direction.y < 0) {
            origin.y += headLength + 1
        }
        return new ArrowHelper(direction, origin, headLength + 1, ARROW_COLOR, headLength, headWidth)
    }
}
