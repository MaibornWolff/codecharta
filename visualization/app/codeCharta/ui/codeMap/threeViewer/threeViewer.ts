import {ThreeViewerService} from "./threeViewerService";
import {ThreeSceneService} from "./threeSceneService";
import {ThreeCameraService} from "./threeCameraService";
import {ThreeOrbitControlsService} from "./threeOrbitControlsService";
import {ThreeRendererService} from "./threeRendererService";
import {ThreeUpdateCycleService} from "./threeUpdateCycleService";
import "../../../core/settings/settings.module";

import angular from "angular";

angular.module("app.codeCharta.ui.codeMap.threeViewer", ["app.codeCharta.core.settings"])
    .service(
        ThreeViewerService.SELECTOR,
        ThreeViewerService
    )
    .service(
        ThreeUpdateCycleService.SELECTOR,
        ThreeUpdateCycleService
    )
    .service(
        ThreeSceneService.SELECTOR,
        ThreeSceneService
    )
    .service(
        ThreeRendererService.SELECTOR,
        ThreeRendererService
    )
    .service(
        ThreeOrbitControlsService.SELECTOR,
        ThreeOrbitControlsService
    )
    .service(
        ThreeCameraService.SELECTOR,
        ThreeCameraService
    );


