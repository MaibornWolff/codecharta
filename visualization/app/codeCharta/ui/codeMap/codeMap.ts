import angular from "angular";

import "./threeViewer/threeViewer";
import "../../core/core.module";

import {codeMapComponent} from "./codeMap.component";
import {CodeMapMouseEventService} from "./codeMap.mouseEvent.service";
import {CodeMapRenderService} from "./codeMap.render.service";
import {CodeMapActionsService} from "./codeMap.actions.service";
import {CodeMapUtilService} from "./codeMap.util.service";
import {CodeMapLabelService} from "./codeMap.label.service";
import {CodeMapArrowService} from "./codeMap.arrow.service";

angular.module("app.codeCharta.ui.codeMap",["app.codeCharta.ui.codeMap.threeViewer", "app.codeCharta.core"])
.component(
    codeMapComponent.selector,
    codeMapComponent
).service(
    CodeMapRenderService.SELECTOR,
    CodeMapRenderService
).service(
    CodeMapMouseEventService.SELECTOR,
    CodeMapMouseEventService
).service(
    CodeMapActionsService.SELECTOR,
    CodeMapActionsService
).service(
    CodeMapUtilService.SELECTOR,
    CodeMapUtilService
).service(
    CodeMapLabelService.SELECTOR,
    CodeMapLabelService
).service(
    CodeMapArrowService.SELECTOR,
    CodeMapArrowService
);