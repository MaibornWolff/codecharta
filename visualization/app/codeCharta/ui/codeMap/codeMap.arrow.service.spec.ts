import "../../codeCharta.module"
import "./codeMap.module"
import { CodeMapArrowService } from "./codeMap.arrow.service"
import { ThreeSceneService } from "./threeViewer/threeSceneService"
import { getService, instantiateModule } from "../../../../mocks/ng.mockhelper"
import { Object3D, Vector3 } from "three"
import { CODE_MAP_BUILDING, OUTGOING_NODE, INCOMING_NODE } from "../../util/dataMocks"
import { Edge, EdgeVisibility, Node } from "../../codeCharta.model"
import { IRootScopeService } from "angular"
import { CodeMapMouseEventService } from "./codeMap.mouseEvent.service"
import { ColorConverter } from "../../util/color/colorConverter"
import { StoreService } from "../../state/store.service"
import { setScaling } from "../../state/store/appSettings/scaling/scaling.actions"

describe("CodeMapArrowService", () => {
	let codeMapArrowService: CodeMapArrowService
	let threeSceneService: ThreeSceneService
	let $rootScope: IRootScopeService
	let storeService: StoreService

	beforeEach(() => {
		restartSystem()
		rebuildService()
		withMockedThreeSceneService()
	})

	afterEach(() => {
		jest.resetAllMocks()
	})

	function restartSystem() {
		instantiateModule("app.codeCharta.ui.codeMap")

		threeSceneService = getService<ThreeSceneService>("threeSceneService")
		storeService = getService<StoreService>("storeService")
		$rootScope = getService<IRootScopeService>("$rootScope")
	}

	function rebuildService() {
		codeMapArrowService = new CodeMapArrowService($rootScope, storeService, threeSceneService)
	}

	function withMockedThreeSceneService() {
		threeSceneService = codeMapArrowService["threeSceneService"] = jest.fn().mockReturnValue({
			edgeArrows: {
				children: [],
				add: jest.fn()
			},
			highlightedBuildings: [],
			getMapMesh: jest.fn().mockReturnValue({
				getMeshDescription: jest.fn().mockReturnValue({ getBuildingByPath: jest.fn() }),
				clearHighlight: jest.fn(),
				highlightBuilding: jest.fn(),
				clearSelection: jest.fn(),
				selectBuilding: jest.fn()
			}),
			addBuildingToHighlightingList: jest.fn(),
			getSelectedBuilding: jest.fn().mockReturnValue({
				value: "value"
			})
		})()
	}

	function setupEdgeArrowsWithChildren() {
		const dummyObject3D = new Object3D()
		threeSceneService.edgeArrows.children = [dummyObject3D, dummyObject3D]
	}

	function setupArrows() {
		const dummyObject3D = new Object3D()
		codeMapArrowService["arrows"] = [dummyObject3D, dummyObject3D]
	}

	describe("constructor", () => {
		it("should assign arrows an empty array", () => {
			expect(codeMapArrowService["arrows"].length).toBe(0)
		})

		it("should subscribe to Building-Hovered-Events", () => {
			CodeMapMouseEventService.subscribeToBuildingHovered = jest.fn()

			rebuildService()

			expect(CodeMapMouseEventService.subscribeToBuildingHovered).toHaveBeenCalledWith($rootScope, codeMapArrowService)
		})

		it("should subscribe to Building-Unhovered-Events", () => {
			CodeMapMouseEventService.subscribeToBuildingUnhovered = jest.fn()

			rebuildService()

			expect(CodeMapMouseEventService.subscribeToBuildingUnhovered).toHaveBeenCalledWith($rootScope, codeMapArrowService)
		})
	})

	describe("SelectionMethods", () => {
		beforeEach(() => {
			codeMapArrowService.clearArrows = jest.fn()
			codeMapArrowService["showEdgesOfBuildings"] = jest.fn()
			codeMapArrowService.addEdgePreview = jest.fn()
		})
		it("should call clearArrows and showEdgesOfBuildings through BuildingSelected", () => {
			codeMapArrowService.onBuildingSelected(CODE_MAP_BUILDING)

			expect(codeMapArrowService.clearArrows).toHaveBeenCalled()
			expect(codeMapArrowService["showEdgesOfBuildings"]).toHaveBeenCalled()
			expect(codeMapArrowService.addEdgePreview).toHaveBeenCalledTimes(0)
		})

		it("should call clearArrows and showEdgesOfBuildings through BuildingHovered", () => {
			codeMapArrowService.onBuildingHovered(CODE_MAP_BUILDING)

			expect(codeMapArrowService.clearArrows).toHaveBeenCalled()
			expect(codeMapArrowService["showEdgesOfBuildings"]).toHaveBeenCalled()
			expect(codeMapArrowService.addEdgePreview).toHaveBeenCalledTimes(0)
		})

		it("should call clearArrows and showEdgesOfBuildings through BuildingUnHovered", () => {
			codeMapArrowService.onBuildingUnhovered()

			expect(codeMapArrowService.clearArrows).toHaveBeenCalled()
			expect(codeMapArrowService["showEdgesOfBuildings"]).toHaveBeenCalled()
			expect(codeMapArrowService.addEdgePreview).toHaveBeenCalledTimes(0)
		})

		it("should call clearArrows and showEdgesOfBuildings through BuildingDeselcted", () => {
			codeMapArrowService.onBuildingDeselected()

			expect(codeMapArrowService.clearArrows).toHaveBeenCalled()
			expect(codeMapArrowService["showEdgesOfBuildings"]).toHaveBeenCalledTimes(0)
			expect(codeMapArrowService.addEdgePreview).toHaveBeenCalled()
		})
	})

	describe("clearArrows", () => {
		it("should remove all array entries of field arrows", () => {
			setupEdgeArrowsWithChildren()

			codeMapArrowService.clearArrows()

			expect(codeMapArrowService["arrows"].length).toBe(0)
		})

		it("should remove all array entries of threeSceneService.edgeArrows.children", () => {
			setupEdgeArrowsWithChildren()

			codeMapArrowService.clearArrows()

			expect(threeSceneService.edgeArrows.children.length).toBe(0)
		})
	})

	describe("addEdgePreview", () => {
		beforeEach(() => {
			codeMapArrowService["map"] = new Map<String, Node>()
			codeMapArrowService["map"].get = jest.fn(() => {
				return INCOMING_NODE
			})
			codeMapArrowService["previewMode"] = jest.fn()
		})
		it("should create and edge Preview of one", () => {
			const nodes: Node[] = [OUTGOING_NODE]
			const edges: Edge[] = storeService.getState().fileSettings.edges.filter(x => x.visible != EdgeVisibility.none)

			codeMapArrowService.addEdgePreview(nodes, edges)

			expect(codeMapArrowService["map"].size).toEqual(1)
		})
		it("should create and no edge Preview at all", () => {
			const edges: Edge[] = storeService.getState().fileSettings.edges.filter(x => x.visible != EdgeVisibility.none)

			codeMapArrowService.addEdgePreview(null, edges)

			expect(codeMapArrowService["map"].size).toEqual(0)
		})
	})

	describe("createCurve", () => {
		it("should create a curve out of the 2 Nodes", () => {
			const originNode: Node = OUTGOING_NODE
			const targetNode: Node = INCOMING_NODE
			const curveScale = 100 * storeService.getState().appSettings.edgeHeight

			const curve = codeMapArrowService["createCurve"](originNode, targetNode, curveScale)

			expect(curve).toBeDefined()
		})
	})

	describe("highlightBuilding", () => {
		it("should highlight certain buildings", () => {
			codeMapArrowService["highlightBuilding"](OUTGOING_NODE)
			expect(threeSceneService.getMapMesh().getMeshDescription().getBuildingByPath).toHaveBeenCalled()
			expect(threeSceneService.addBuildingToHighlightingList).toHaveBeenCalled()
		})
	})

	describe("setCurveColor", () => {
		beforeEach(() => {
			threeSceneService.edgeArrows["add"] = jest.fn()
			codeMapArrowService["arrows"].push = jest.fn()
		})
		it("should run through the function with mocked subfunctions", () => {
			const originNode: Node = OUTGOING_NODE
			const targetNode: Node = INCOMING_NODE
			const curveScale = 100 * storeService.getState().appSettings.edgeHeight
			const curve = codeMapArrowService["createCurve"](originNode, targetNode, curveScale)
			const color = ColorConverter.convertHexToNumber(storeService.getState().appSettings.mapColors.outgoingEdge)

			codeMapArrowService["setCurveColor"](curve, color)

			expect(threeSceneService.edgeArrows["add"]).toHaveBeenCalled()
			expect(codeMapArrowService["arrows"].push).toHaveBeenCalled()
		})
	})

	describe("scale", () => {
		it("should set the scale of all arrows to x, y and z", () => {
			setupArrows()
			storeService.dispatch(setScaling(new Vector3(1, 2, 3)))

			codeMapArrowService.scale()

			expect(codeMapArrowService["arrows"][0].scale.x).toBe(1)
			expect(codeMapArrowService["arrows"][0].scale.y).toBe(2)
			expect(codeMapArrowService["arrows"][0].scale.z).toBe(3)
		})
	})
})
