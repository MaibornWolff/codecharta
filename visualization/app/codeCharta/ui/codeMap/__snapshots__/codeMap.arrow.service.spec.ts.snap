// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeMapArrowService Arrow Behaviour when selecting and hovering a building should only highlight small leaf when big leaf is selected 1`] = `
Array [
  CodeMapBuilding {
    "_boundingBox": Box3 {
      "max": Vector3 {
        "x": 6,
        "y": 8,
        "z": 10,
      },
      "min": Vector3 {
        "x": 5,
        "y": 6,
        "z": 7,
      },
    },
    "_color": "#AABBCC",
    "_defaultColor": "#AABBCC",
    "_defaultDeltaColor": "#000000",
    "_deltaColor": "#000000",
    "_id": 1,
    "_node": Object {
      "attributes": Object {
        "a": 20,
        "b": 15,
        "mcc": 14,
      },
      "color": "#AABBCC",
      "deltas": Object {
        "a": 1,
        "b": 2,
      },
      "depth": 4,
      "edgeAttributes": Object {
        "a": Object {
          "incoming": 2,
          "outgoing": 666,
        },
      },
      "flat": false,
      "height": 2,
      "heightDelta": 20,
      "id": 2,
      "incomingEdgePoint": Vector3 {
        "x": 1,
        "y": 2,
        "z": 3,
      },
      "isLeaf": true,
      "length": 3,
      "link": "NO_LINK",
      "markingColor": "0xFFFFFF",
      "name": "root/small leaf",
      "outgoingEdgePoint": Vector3 {
        "x": 1,
        "y": 2,
        "z": 3,
      },
      "path": "/root/Parent Leaf/small leaf",
      "visible": true,
      "width": 1,
      "x0": 5,
      "y0": 7,
      "z0": 6,
    },
  },
  CodeMapBuilding {
    "_boundingBox": Box3 {
      "max": Vector3 {
        "x": 6,
        "y": 8,
        "z": 10,
      },
      "min": Vector3 {
        "x": 5,
        "y": 6,
        "z": 7,
      },
    },
    "_color": "#AABBCC",
    "_defaultColor": "#AABBCC",
    "_defaultDeltaColor": "#000000",
    "_deltaColor": "#000000",
    "_id": 1,
    "_node": Object {
      "attributes": Object {
        "a": 20,
        "b": 15,
        "mcc": 14,
      },
      "color": "#AABBCC",
      "deltas": Object {
        "a": 1,
        "b": 2,
      },
      "depth": 4,
      "edgeAttributes": Object {
        "a": Object {
          "incoming": 2,
          "outgoing": 666,
        },
      },
      "flat": false,
      "height": 2,
      "heightDelta": 20,
      "id": 2,
      "incomingEdgePoint": Vector3 {
        "x": 1,
        "y": 2,
        "z": 3,
      },
      "isLeaf": true,
      "length": 3,
      "link": "NO_LINK",
      "markingColor": "0xFFFFFF",
      "name": "root/small leaf",
      "outgoingEdgePoint": Vector3 {
        "x": 1,
        "y": 2,
        "z": 3,
      },
      "path": "/root/Parent Leaf/small leaf",
      "visible": true,
      "width": 1,
      "x0": 5,
      "y0": 7,
      "z0": 6,
    },
  },
]
`;

exports[`CodeMapArrowService Arrow Behaviour when selecting and hovering a building should only highlight small leaf when big leaf is selected 2`] = `
CodeMapBuilding {
  "_boundingBox": Box3 {
    "max": Vector3 {
      "x": -Infinity,
      "y": -Infinity,
      "z": -Infinity,
    },
    "min": Vector3 {
      "x": Infinity,
      "y": Infinity,
      "z": Infinity,
    },
  },
  "_color": "#EB8319",
  "_defaultColor": "#ddcc00",
  "_defaultDeltaColor": "#000000",
  "_deltaColor": "#000000",
  "_id": 1,
  "_node": Object {
    "attributes": Object {
      "a": 20,
      "b": 15,
      "mcc": 14,
    },
    "color": "#AABBCC",
    "deltas": Object {
      "a": 1,
      "b": 2,
    },
    "depth": 4,
    "edgeAttributes": Object {
      "a": Object {
        "incoming": 2,
        "outgoing": 666,
      },
    },
    "flat": false,
    "height": 2,
    "heightDelta": 20,
    "id": 1,
    "incomingEdgePoint": Vector3 {
      "x": 1,
      "y": 2,
      "z": 3,
    },
    "isLeaf": true,
    "length": 3,
    "link": "NO_LINK",
    "markingColor": "0xFFFFFF",
    "name": "root/big leaf",
    "outgoingEdgePoint": Vector3 {
      "x": 1,
      "y": 2,
      "z": 3,
    },
    "path": "/root/big leaf",
    "visible": true,
    "width": 1,
    "x0": 5,
    "y0": 7,
    "z0": 6,
  },
}
`;
